implementations/rust/ockam/ockam/tests/forwarder.rs

```diff
@@ -0,0 +1,165 @@
+use ockam::remote::{RemoteForwarder, RemoteForwarderTrustOptions};
+use ockam::workers::Echoer;
+use ockam::ForwardingService;
+use ockam_core::sessions::{SessionPolicy, Sessions};
+use ockam_core::{route, Address, AllowAll, Result};
+use ockam_node::{Context, MessageReceiveOptions, MessageSendReceiveOptions};
+use ockam_transport_tcp::{TcpConnectionTrustOptions, TcpListenerTrustOptions, TcpTransport};
+use std::time::Duration;
+
+// Node creates a Forwarding service and a Remote Forwarder, Echoer is reached through the Forwarder. No session
+#[ockam_macros::test]
+async fn test1(ctx: &mut Context) -> Result<()> {
+    ForwardingService::create(ctx, "forwarding_service", AllowAll, AllowAll).await?;
+
+    ctx.start_worker("echoer", Echoer, AllowAll, AllowAll)
+        .await?;
+
+    let remote_info =
+        RemoteForwarder::create(ctx, route![], RemoteForwarderTrustOptions::new()).await?;
+
+    let resp = ctx
+        .send_and_receive_extended::<String>(
+            route![remote_info.remote_address(), "echoer"],
+            "Hello".to_string(),
+            MessageSendReceiveOptions::new(),
+        )
+        .await?;
+
+    assert_eq!(resp, "Hello");
+
+    ctx.stop().await
+}
+
+// Cloud: Hosts a Forwarding service and listens on a tcp port. No session
+// Server: Connects to a Cloud using tcp and creates a dynamic Forwarder. Using session
+// Client: Connects to a Cloud using tcp and reaches to the Server's Echoer. Using session
+#[ockam_macros::test]
+async fn test2(ctx: &mut Context) -> Result<()> {
+    ForwardingService::create(ctx, "forwarding_service", AllowAll, AllowAll).await?;
+    let cloud_tcp = TcpTransport::create(ctx).await?;
+    let (socket_addr, _) = cloud_tcp
+        .listen("127.0.0.1:0", TcpListenerTrustOptions::new())
+        .await?;
+
+    let server_sessions = Sessions::default();
+    let server_tcp_session_id = server_sessions.generate_session_id();
+
+    ctx.start_worker("echoer", Echoer, AllowAll, AllowAll)
+        .await?;
+    server_sessions.add_consumer(
+        &Address::from_string("echoer"),
+        &server_tcp_session_id,
+        SessionPolicy::ProducerAllowMultiple,
+    );
+
+    let server_tcp = TcpTransport::create(ctx).await?;
+    let cloud_connection = server_tcp
+        .connect(
+            socket_addr.to_string(),
+            TcpConnectionTrustOptions::new().as_producer(&server_sessions, &server_tcp_session_id),
+        )
+        .await?;
+
+    let remote_info = RemoteForwarder::create(
```

Comment:
I think it would be nice to have an additional test where we create a Forwarder via a secure channel.